overlord/snapstate/handlers.go

```diff
@@ -824,3 +824,216 @@ func (m *SnapManager) doDiscardSnap(t *state.Task, _ *tomb.Tomb) error {
+	newAliases, nAuto, err := refreshAliases(st, curInfo, curAliases)
+	if err != nil {
+		return err
+	}
+	if status.NoAutoYet() && nAuto != 0 {
+		if !status.Pending() {
+			status = EnabledAliases
+		} else {
+			status = PendingEnabledAliases
+		}
+	}
+	_, err = checkAliasesConflicts(st, snapName, status, newAliases)
+	if err != nil {
+		return err
+	}
+
+	if !status.Pending() {
+		if err := applyAliasesChange(st, snapName, status, curAliases, status, newAliases, m.backend); err != nil {
+			return err
+		}
+	}
+
+	t.Set("old-aliases-v2", curAliases)
+	t.Set("old-aliases-v2-status", snapst.AliasesStatus)
+	snapst.AliasesStatus = status
+	snapst.Aliases = newAliases
+	Set(st, snapName, snapst)
+	return nil
+}
+
+func (m *SnapManager) undoRefreshAliasesV2(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	var oldAliases map[string]*AliasTarget
+	err := t.Get("old-aliases-v2", &oldAliases)
+	if err == state.ErrNoState {
+		// nothing to do
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+	var oldStatus AliasesStatus
+	if err := t.Get("old-aliases-v2-status", &oldStatus); err != nil {
+		return err
+	}
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.Name()
+	curStatus := snapst.AliasesStatus
+	status := curStatus
+	if oldStatus.NoAutoYet() {
+		// old state represented not having any auto aliases yet at all
+		status = oldStatus
+	}
+
+	// check if the old states creates conflicts now
+	_, err = checkAliasesConflicts(st, snapName, status, oldAliases)
+	if _, ok := err.(*AliasConflictError); ok {
+		// best we can do is reinstate with all aliases disabled
+		t.Errorf("cannot reinstate alias state because of conflicts, disabling: %v", err)
+		status, oldAliases = disableAliases(status, oldAliases)
+	} else if err != nil {
+		return err
+	}
+
+	if t.Kind() != "set-auto-aliases-v2" && !status.Pending() {
+		curAliases := snapst.Aliases
+		if err := applyAliasesChange(st, snapName, curStatus, curAliases, status, oldAliases, m.backend); err != nil {
+			return err
+		}
+	}
+
+	snapst.AliasesStatus = status
+	snapst.Aliases = oldAliases
+	Set(st, snapName, snapst)
+	return nil
+}
+
+func (m *SnapManager) doPruneAutoAliasesV2(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+	snapsup, snapst, err := snapSetupAndState(t)
+	if err != nil {
+		return err
+	}
+	var which []string
+	err = t.Get("aliases", &which)
+	if err != nil {
+		return err
+	}
+	snapName := snapsup.Name()
+	status := snapst.AliasesStatus
+	curAliases := snapst.Aliases
+
+	newAliases := pruneAutoAliases(st, curAliases, which)
+
+	if !status.Pending() {
+		if err := applyAliasesChange(st, snapName, status, curAliases, status, newAliases, m.backend); err != nil {
+			return err
+		}
+	}
+
+	t.Set("old-aliases-v2", curAliases)
+	t.Set("old-aliases-v2-status", snapst.AliasesStatus)
+	snapst.Aliases = newAliases
+	Set(st, snapName, snapst)
+	return nil
+}
```

Comment:
These task handlers are super readable despite the complexity of the problem.

Thanks for pulling that off!